runtime/ops/tls.rs

```diff
@@ -70,6 +88,457 @@ impl StoresClientSessions for ClientSessionMemoryCache {
   }
 }
 
+enum TlsSession {
+  Client(ClientSession),
+  Server(ServerSession),
+}
+
+impl Deref for TlsSession {
+  type Target = dyn Session;
+
+  fn deref(&self) -> &Self::Target {
+    match self {
+      TlsSession::Client(client_session) => client_session,
+      TlsSession::Server(server_session) => server_session,
+    }
+  }
+}
+
+impl DerefMut for TlsSession {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    match self {
+      TlsSession::Client(client_session) => client_session,
+      TlsSession::Server(server_session) => server_session,
+    }
+  }
+}
+
+impl From<ClientSession> for TlsSession {
+  fn from(client_session: ClientSession) -> Self {
+    TlsSession::Client(client_session)
+  }
+}
+
+impl From<ServerSession> for TlsSession {
+  fn from(server_session: ServerSession) -> Self {
+    TlsSession::Server(server_session)
+  }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Flow {
+  Read,
+  Write,
+}
+
+#[allow(clippy::enum_variant_names)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Shut {
+  Open,
+  ExpectEof,
+  TlsShut,
+  TcpShut,
+}
+
+pub struct TlsStream {
+  tls: TlsSession,
+  tcp: TcpStream,
+  rd_shut: Shut,
+  wr_shut: Shut,
+}
+
+impl TlsStream {
+  fn new(tcp: TcpStream, tls: TlsSession) -> Self {
+    Self {
+      tcp,
+      tls,
+      rd_shut: Shut::Open,
+      wr_shut: Shut::Open,
+    }
+  }
+
+  fn new_client_side(
+    tcp: TcpStream,
+    tls_config: &Arc<ClientConfig>,
+    hostname: DNSNameRef,
+  ) -> Self {
+    let tls = TlsSession::Client(ClientSession::new(tls_config, hostname));
+    Self::new(tcp, tls)
+  }
+
+  fn new_server_side(tcp: TcpStream, tls_config: &Arc<ServerConfig>) -> Self {
+    let tls = TlsSession::Server(ServerSession::new(tls_config));
+    Self::new(tcp, tls)
+  }
+
+  fn into_split(self) -> (ReadHalf, WriteHalf) {
+    let shared = Shared::new(self);
+    let rd = ReadHalf {
+      shared: shared.clone(),
+    };
+    let wr = WriteHalf { shared };
+    (rd, wr)
+  }
+
+  fn poll_io(
```

Comment:
Please note that it is dangerous to use the same `poll_io` for reading and writing. see https://github.com/tokio-rs/tls/issues/40

Reply:
In what way is it dangerous? Note that tokio-rs/tls#40 is the reason we're migrating away from tokio-rustls.

A special `Waker` implementation (see https://github.com/denoland/deno/blob/43beea27809e7a29560dd29ecbd18ca38a0fa133/runtime/ops/tls.rs#L445-L461) makes it safe to read the underlying stream when writing, as well as writing to the underlying stream when reading.

Reply:
You may have misunderstood something, `tokio-rustls` does not have the problem described in https://github.com/tokio-rs/tls/issues/40, because `tokio-rustls` does not perform reverse IO, you can see it in [the comment](https://github.com/tokio-rs/tls/issues/40#issuecomment-735037105).

Shared looks interesting, but I suspect it has performance issues because it may cause unnecessary wakeups. and since `rd_waker` and `wr_waker` always wake up at same time, it may cause `tls_stream` to always have lock contention.

Reply:
I agree that in lock contention could be an issue when using the multi-threaded runtime.
However, Deno is using the single threaded runtime, so that's not going to be a problem for us.

I necessary, I could imagine using a different scheme (for example, we could daisy chain the wake-ups: one task wakes up another waker when polled). But I don't think we need that extra complexity right now.

Reply:
Fair enough, I don't care what implementation deno uses. I am glad to know that `tokio-rustls` has no problems.

I hope you can edit the description of PR to avoid unnecessary misunderstandings.